Rearranging support in playlist

kOchirasu opened the pull request:
Should be working, but only thing I feel is not good is that there should probably be a line showing where the dragged track will be inserted. Not sure if there is a design I should follow for that so I didn't add it.

Closes #810.

kOchirasu committed:
Allow rearranging playlist tracks

kOchirasu committed:
Use PlaylistList to manage Prev/Next tracks

kOchirasu committed:
Fixed code style errors

2yangk23 commented:
nvm I didn't even realize other people already implemented this.

smoogipooo commented on osu.Game/Overlays/Music/PlaylistItem.cs:
```diff
@@ -114,24 +117,62 @@ private void recreateText()
+        protected override bool OnDragEnd(InputState state)
+        {
+            int src = (int) Depth;
+            int dst = clamp((int) ((state.Mouse.Position.Y - Parent.DrawPosition.Y) / Height), 0, playlist.Count - 1);
```
`MathHelper.Clamp`?

smoogipooo replied:
And no spacing between cast and the variables.

smoogipooo replied:
Also DrawPosition in this case isn't needed because Mouse.Position.Y is already relative to us.

2yangk23 replied:
I'm pretty sure the DrawPosition is not 0 and this works, so removing it wouldn't work, but I will try it without to make sure.

2yangk23 replied:
Turns out the DrawPosition is 0, so removed.

smoogipooo commented on osu.Game/Overlays/Music/PlaylistItem.cs:
```diff
@@ -114,24 +117,62 @@ private void recreateText()
+            {
+                for (int i = dst; i < src; i++)
+                    playlist.ChangeChildDepth(playlist[i], i + 1);
+            }
```
Add line break here.

smoogipooo commented on osu.Game/Overlays/Music/PlaylistItem.cs:
```diff
@@ -114,24 +117,62 @@ private void recreateText()
+            return true;
+        }
+
+        protected override bool OnDragEnd(InputState state)
```
How about moving this code to OnDrag instead?

2yangk23 replied:
It only needs to be computed once. From what I saw OnDrag is constantly being invoked while the object is being dragged.

smoogipooo replied:
That's exactly what I'd expect though - instantaneous feedback of some sort. Otherwise you need to add the "ghost text" or whatever you had in mind.

I personally had in mind the item popping out and providing a blank space, however I would've been happy with the effect this causes if it's recomputed in OnDrag.

The computation itself is in-expensive and has the correct src == dst check.

2yangk23 replied:
Now that I think about it, it is inexpensive. When I was initially working on it, I was thinking that the computation is not very cheap since you have to shift all affected tracks. However, realistically you only move it by 1 track at a time so at most it's affecting two tracks. I will make the change.

smoogipoo reviewed (changes requested):
Minor review for now - more is coming.

kOchirasu committed:
Use MathHelper.Clamp

- Fix formatting

smoogipooo commented on osu.Game/Overlays/Music/PlaylistItem.cs:
```diff
@@ -114,24 +117,58 @@ private void recreateText()
 
+        protected override bool OnDragStart(InputState state)
+        {
+            return true;
```
The items should only be draggable if the little cheeseburger thing to the left of the items is hovered.

kOchirasu committed:
Rearrange tracks in OnDrag

smoogipooo commented on osu.Game/Overlays/Music/PlaylistItem.cs:
```diff
@@ -114,24 +117,58 @@ private void recreateText()
+        // Maybe render some ghost text
+        protected override bool OnDrag(InputState state)
+        {
+            return true;
```
All of these that have a one line can just be, e.g. `OnDrag(InputState state) => true;`

smoogipooo commented on osu.Game/Overlays/Music/PlaylistItem.cs:
```diff
@@ -114,24 +117,58 @@ private void recreateText()
+        protected override bool OnDragEnd(InputState state)
+        {
+            int src = (int)Depth;
+            int dst = MathHelper.Clamp((int)(state.Mouse.Position.Y / Height), 0, playlist.Count - 1);
```
The calculation here seems to give an off-by-one after scrolling down the list and dragging. Haven't found the cause just yet, but... https://streamable.com/mgjty

2yangk23 replied:
I can't reproduce this, but I only have maybe 1.5 pages of songs. Does this still happen for you if you only scroll down a little?

smoogipooo replied:
I think I know the cause of this - items that span multiple lines:
![image](https://user-images.githubusercontent.com/1329837/29806787-150e71c4-8ccc-11e7-95e0-014e26c3da96.png)

2yangk23 replied:
rip, that breaks everything. Is this a valid usecase? Is there a plan to use ellipses.

smoogipooo replied:
Don't think so, but I don't think this should be using mouse position anyway in this case. How about doing something like `if (GetContainingInputManager().DraggedDrawable != this) { /* replace dragged drawable */ }`?

kOchirasu committed:
Rearrange dragging using cheeseburger icon only

kOchirasu committed:
Get destination index using binarysearch

kOchirasu committed:
Fix out of bounds

2yangk23 commented:
I feel it stutters a bit when dragging slowly so it still needs some refinement. Maybe I'll look into it tomorrow.

kOchirasu committed:
Try to fix stuttering behavior

2yangk23 commented:
I wouldn't call it the most elegant solution, but I tried to fix the stutter. I think it is much better than before, but it can still happens sometimes. I can't really explain why it still happens though.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into pr/n1202_2yangk23

smoogipoo committed:
Fix items jumping between two indices in the edge case, use a linear search for now.

smoogipoo committed:
Use TakeWhile instead of reversing the list.

smoogipoo committed:
Rewrite PlaylistList as CompositeDrawable and remove all backwards PlaylistList references

Now handles drag at a PlaylistList.ItemsScrollContainer level (private class), and PlaylistList itself is no longer a Container so it only supports adding BeatmapSets. Sorry for the rewrite x.x.

smoogipoo committed:
Add better fix for items swapping erratically.

smoogipoo committed:
Add scrolling points so items can be dragged beyond the list.

smoogipoo committed:
Clamp at the end points.

smoogipoo committed:
Fix CI error.

smoogipooo commented:
@peppy requesting a review on this as I've pretty much rewritten it (apologies, @2yangk23 ). I've implemented IDraggable for now, with the expectation that at some point in the future we'll move all of this code into a "DragSupportingScrollContainer" or something.

peppy commented on osu.Game/Overlays/Music/PlaylistList.cs:
```diff
@@ -4,104 +4,249 @@
+                if (draggedItem == null)
+                    return;
+
+                var mouseState = GetContainingInputManager().CurrentState.Mouse;
```
we can't

smoogipoo committed:
Don't query GetContainingInputManager every Update.

smoogipoo committed:
Merge branch 'master' into rearrange-playlist

smoogipooo commented:
I hope this is less abusive?

peppy commented on osu.Game/Overlays/Music/PlaylistList.cs:
```diff
@@ -127,21 +125,27 @@ public BeatmapSetInfo SelectedSet
             public BeatmapSetInfo NextSet => (items.SkipWhile(i => !i.Selected).Skip(1).FirstOrDefault() ?? items.FirstOrDefault())?.BeatmapSetInfo;
             public BeatmapSetInfo PreviousSet => (items.TakeWhile(i => !i.Selected).LastOrDefault() ?? items.LastOrDefault())?.BeatmapSetInfo;
 
+            private InputState dragInputState;
```
you could potentially just store the Vector2 here

peppy commented on osu.Game/Overlays/Music/PlaylistList.cs:
```diff
@@ -127,21 +125,27 @@ public BeatmapSetInfo SelectedSet
             protected override bool OnDrag(InputState state)
             {
+                dragInputState = state;
                 if (draggedItem == null)
```
pretty sure this check is not required (`OnDrag` is only sent if returning true to `OnDragStart`)

smoogipooo replied:
and returning true in `OnDragStart` does not necessarily imply that the little handle to the left of the items was dragged.

peppy replied:
why not? i could understand if we were a `ScrollContainer`, but how does this work?

smoogipooo replied:
we are a `ScrollContainer`

peppy replied:
nested classes sure don't diff well

smoogipoo committed:
Only store native position instead of input state.

peppy commented:
todo: move implementation to framework at some later point in time
